packages/std/src/testing/mock.rs

```diff
@@ -142,62 +122,31 @@ impl Api for MockApi {
-        // Dummy input validation. This is more sophisticated for formats like bech32, where format and checksum are validated.
-        let min_length = 3;
-        let max_length = self.canonical_length;
-        if input.len() < min_length {
-            return Err(StdError::generic_err(
-                format!("Invalid input: human address too short for this mock implementation (must be >= {min_length})."),
-            ));
-        }
-        if input.len() > max_length {
-            return Err(StdError::generic_err(
-                format!("Invalid input: human address too long for this mock implementation (must be <= {max_length})."),
-            ));
-        }
-
-        // mimics formats like hex or bech32 where different casings are valid for one address
-        let normalized = input.to_lowercase();
-
-        let mut out = Vec::from(normalized);
-
-        // pad to canonical length with NULL bytes
-        out.resize(self.canonical_length, 0x00);
-        // content-dependent rotate followed by shuffle to destroy
-        // the most obvious structure (https://github.com/CosmWasm/cosmwasm/issues/552)
-        let rotate_by = digit_sum(&out) % self.canonical_length;
-        out.rotate_left(rotate_by);
-        for _ in 0..SHUFFLES_ENCODE {
-            out = riffle_shuffle(&out);
+        if let Ok((prefix, decoded, Variant::Bech32)) = decode(input) {
+            if prefix == self.bech32_prefix {
+                if let Ok(bytes) = Vec::<u8>::from_base32(&decoded) {
+                    validate_length(&bytes)?;
+                    return Ok(bytes.into());
+                }
+            }
         }
-        Ok(out.into())
+        Err(StdError::generic_err("Invalid input"))
     }
 
     fn addr_humanize(&self, canonical: &CanonicalAddr) -> StdResult<Addr> {
-        if canonical.len() != self.canonical_length {
-            return Err(StdError::generic_err(
-                "Invalid input: canonical address length not correct",
-            ));
-        }
+        validate_length(canonical.as_ref())?;
 
-        let mut tmp: Vec<u8> = canonical.clone().into();
-        // Shuffle two more times which restored the original value (24 elements are back to original after 20 rounds)
-        for _ in 0..SHUFFLES_DECODE {
-            tmp = riffle_shuffle(&tmp);
-        }
-        // Rotate back
-        let rotate_by = digit_sum(&tmp) % self.canonical_length;
-        tmp.rotate_right(rotate_by);
-        // Remove NULL bytes (i.e. the padding)
-        let trimmed = tmp.into_iter().filter(|&x| x != 0x00).collect();
-        // decode UTF-8 bytes into string
-        let human = String::from_utf8(trimmed)?;
-        Ok(Addr::unchecked(human))
+        encode(
+            self.bech32_prefix,
+            canonical.as_slice().to_base32(),
+            Variant::Bech32,
+        )
+        .map(Addr::unchecked)
+        .map_err(|_| StdError::generic_err("Invalid canonical address"))
```

Comment:
Since we already checked the canonical address bytes length, I don't see how this can happen. Maybe turn it into `panic!("Unexpected bech32 encoding error")`?

Reply:
It has nothing to do with the bytes length. It errors for invalid human-readable parts. I'll adjust the error message to reflect that